### PR TITLE
Issue #SB-17658- Migrate Content Progress metrics to a data product 

### PR DESCRIPTION
**Pull request by utk14**

### CODE REVIEW

File: data-products/src/main/scala/org/sunbird/analytics/model/report/TextbookProgressModel.scala [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+      val aggregatedReportRDD = getAggregatedReport(contentData, slugName)
+      val liveStatusReportRDD = getLiveStatusReport(contentData, slugName)
+      val nonLiveStatusReportRDD = getNonLiveStatusReport(contentData, slugName)
+
+      val aggregatedMappingRDD = aggregatedReportRDD.map(f => (f.identifier,f))
+      val liveStatusMappingRDD = liveStatusReportRDD.map(f => (f.identifier,f))
+      val nonLiveStatusMappingRDD = nonLiveStatusReportRDD.map(f => (f.identifier,f))
+
+      val joinedList_1 = aggregatedMappingRDD.fullOuterJoin(liveStatusMappingRDD)
+      val joinedList_2 = joinedList_1.fullOuterJoin(nonLiveStatusMappingRDD)
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/model/report/TextbookProgressModel.scala]: Why don't u merge these two lines?

File: data-products/src/main/scala/org/sunbird/analytics/util/TextbookUtils.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+                     |        "limit": 10000,
+                     |        "facets": [
+                     |          "status"
+                     |        ]
+                     |      }
+                     |    }""".stripMargin
+
+    val resultData = restUtil.post[ContentInformation](url, request)
+      val finalResult = resultData.result
+    finalResult
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/util/TextbookUtils.scala]: Can we build a retry logic? Why don't we use Unirest here which would internally automatically retry for configured number of times. Reach out to @manju, he has used it in one samza job

File: data-products/src/main/scala/org/sunbird/analytics/model/report/TextbookProgressModel.scala [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+      val nonLiveStatus = data.map(f => f.nonLiveStatusReport.getOrElse(NonLiveStatusReport("","",List(""),"","","","","","","Unknown")))
+      reportConfig.output.map{f =>
+        val aggDF = aggregated.toDF()
+        CourseUtils.postDataToBlob(aggDF, f, config)
+
+        val liveStatusDF = liveStatus.toDF()
+        CourseUtils.postDataToBlob(liveStatusDF, f, config)
+
+        val nonLiveStatusDF = nonLiveStatus.toDF()
+        CourseUtils.postDataToBlob(nonLiveStatusDF, f, config)
```
Comment (SanthoshVasabhaktula) [inline data-products/src/main/scala/org/sunbird/analytics/model/report/TextbookProgressModel.scala]: Can we also get some stats on records count per channel/slug, time taken per channel etc here which would help in debugging?

File: data-products/src/main/scala/org/sunbird/analytics/model/report/TextbookProgressModel.scala [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+case class TenantInfo(id: String, slug: String) extends AlgoInput
+
+case class ContentResult(channel: String,identifier: String, board: String, gradeLevel: List[String], medium: Object, subject: Object,
+                         status: String, creator: String,lastPublishedOn: String,lastSubmittedOn: String, createdFor: List[String],
+                         createdOn: String, contentType: String, mimeType: String,
+                         resourceType: Object, pkgVersion: Integer)
+
+case class AggregatedReport (board: String, medium: Object, gradeLevel: Seq[String], subject: Object, resourceType: Object,
+                             live: Integer, review: Integer, draft: Integer, unlisted: Integer,
+                             application_ecml: Integer, vedio_youtube: Integer, video_mp4: Integer, application_pdf: Integer,
```
[the remark this answers is not in the document]
Comment (utk14) [reply data-products/src/main/scala/org/sunbird/analytics/model/report/TextbookProgressModel.scala]: Done

File: data-products/src/main/scala/org/sunbird/analytics/model/report/TextbookProgressModel.scala [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+                             application_html: Integer,identifier: String, slug: String)
+
+case class LiveReport (board: String, medium: Object,gradeLevel: List[String], identifier: String,resourceType: Object, createdOn: String,pkgVersion: Integer, creator: String,
+                       lastPublishedOn: String, slug: String)
+
+case class NonLiveStatusReport(board: String, medium: Object,gradeLevel: List[String], identifier: String,resourceType: Object, status: String, pendingInCurrentStatus: String,
+                               creator: String, createdOn: String, slug: String)
```
[the remark this answers is not in the document]
Comment (utk14) [reply data-products/src/main/scala/org/sunbird/analytics/model/report/TextbookProgressModel.scala]: Some of the fields required for both the reports are different.

File: data-products/src/main/scala/org/sunbird/analytics/model/report/TextbookProgressModel.scala [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    val data = CourseUtils.getTenantInfo(spark, loadData)
+    val id = data.as[TenantInfo](tenantEncoder).rdd
+    id
+
+  }
+
+  override def algorithm(data: RDD[TenantInfo], config: Map[String, AnyRef])(implicit sc: SparkContext, fc: FrameworkContext): RDD[FinalOutput] = {
+    val result = data.collect().map{f =>
+     val finalResult = getContentData(f.id, f.slug)
+      finalResult
```
[the remark this answers is not in the document]
Comment (utk14) [reply data-products/src/main/scala/org/sunbird/analytics/model/report/TextbookProgressModel.scala]: Done

File: data-products/src/main/scala/org/sunbird/analytics/model/report/TextbookProgressModel.scala (hunk 1 above)
Comment (utk14) [reply data-products/src/main/scala/org/sunbird/analytics/model/report/TextbookProgressModel.scala]: Done

File: data-products/src/main/scala/org/sunbird/analytics/model/report/TextbookProgressModel.scala (hunk 3 above)
Comment (utk14) [reply data-products/src/main/scala/org/sunbird/analytics/model/report/TextbookProgressModel.scala]: Done

File: data-products/src/main/scala/org/sunbird/analytics/util/TextbookUtils.scala (hunk 2 above)
Comment (utk14) [reply data-products/src/main/scala/org/sunbird/analytics/util/TextbookUtils.scala]: Done

File: data-products/src/main/scala/org/sunbird/analytics/model/report/TextbookProgressModel.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (SanthoshVasabhaktula) [reply data-products/src/main/scala/org/sunbird/analytics/model/report/TextbookProgressModel.scala]: Merge them - Have only one case class

**Pull request closed without merging**